### ISSUE TITLE
Visual Studio 2017 hangs, crashes when opening large F# solution

### ISSUE THREAD

**Issue by asik** [Microsoft/visualfsharp#2793]
We have a large (~90 projects) solution, mostly F#. In VS2015 it takes maybe 20-30 seconds to open and the memory usage is around 400MB. In VS2017, the IDE hangs after loading all the projects, memory usage quickly climbs in the 3GB range, and it can stay like that for several minutes before eventually crashing.

This seems related to F# as larger C# solutions don't present this problem.

This is closed-source so of course I cannot share the code, but I would be glad to provide any non-sensitive data to help diagnose this issue. I tried simply creating a new solution and adding our projects to it one by one. Around 60 projects the IDE became extremely slow and usually cannot open the solution anymore. If it manages to get past this loading phase, memory usage goes back to something normal. Then sometimes it becomes usable, sometimes it crashes anyway.

I just tried this again with VS Community 2017 (15.0.0+26228.12) and Visual F# Tools 15.4.1.17040401.

Here's a vmmap screenshot while it's hovering around 3GB:

![image](https://cloud.githubusercontent.com/assets/1631157/24668323/82d36ac2-1934-11e7-8840-0e646c0139cc.png)

**Comment by vasily-kirichenko** [comment]
Would you mine to take a gotMemory snapshot and share it?

**Comment by vasily-kirichenko** [comment]
Do I understand correctly that no files are open?

**Comment by asik** [comment]
Yes I just open Visual Studio 2017 and open the solution.

Not sure about sharing a dump, it's going to have file names and whatnot... but here are some screenshots in dotMemory when devenv.exe was around 1GB and climbing:

![image](https://cloud.githubusercontent.com/assets/1631157/24670101/d251f0c2-193a-11e7-9022-2961221f8e68.png)
![image](https://cloud.githubusercontent.com/assets/1631157/24670158/0dc5abd0-193b-11e7-90e6-ed5fa8bf2224.png)
![image](https://cloud.githubusercontent.com/assets/1631157/24670181/203ccbd6-193b-11e7-9690-57efb2e7e7f4.png)

Going by the vmmap info it might not be managed objects at all?

**Comment by vasily-kirichenko** [comment]
Wow o_O. There should be strictly one `FSharpChecker`

**Comment by vasily-kirichenko** [comment]
Can you show retaining paths for FSharpChecker?

**Comment by asik** [comment]
Oh wait, the number on the right is the number of objects _under_ it. So there are only 2 FsharpChecker objects here (the number on the right).

**Comment by asik** [comment]
As vmmap suggests it looks like the growth in .NET objects is not the problem, it's mainly unmanaged memory:

![image](https://cloud.githubusercontent.com/assets/1631157/24671553/dba5a010-193f-11e7-831b-711d84cca908.png)

FWIW though here are the deltas for .NET objects, sorted by Objects delta:

![image](https://cloud.githubusercontent.com/assets/1631157/24671575/f4c8e084-193f-11e7-9792-81034b93fc31.png)

**Comment by vasily-kirichenko** [comment]
From how many project do you have files opened? Can you open a single file from any project (and restart)?

**Comment by asik** [comment]
So this is all happening before any files are open in the editor. VS is hung on the Start Page, the Solution Explorer is empty. If it manages to load it (I have a smaller one with 61 projects that just barely loads, sometimes), right now one .fs file opens in the editor. The diff from my last post was on this solution.

**Comment by vasily-kirichenko** [comment]
Do you mind try https://drive.google.com/open?id=0B8HLQUKik9VtNFF1UEZNaElhZTg? I reduce project cache size from 200 to 5 in this build. First, open the dev command prompt and run 

`Common7\IDE\VSIXInstaller.exe /uninstall:VisualFSharp`

Then run my vsix.

**Comment by asik** [comment]
Using your VSIX, I get "The 'FSharpProjectPackage' package did not load correctly in VS. In ActivityLog.xml this seems relevant:

`System.IO.FileLoadException: Could not load file or assembly &apos;file:///C:\Users\***\AppData\Local\Microsoft\VisualStudio\15.0_4ee222e8\Extensions\thoyrmux.2tv\FSharp.ProjectSystem.FSharp.dll&apos; or one of its dependencies. Strong name validation failed. (Exception from HRESULT: 0x8013141A)&#x000D;&#x000A;File name: &apos;file:///C:\Users\***\AppData\Local\Microsoft\VisualStudio\15.0_4ee222e8\Extensions\thoyrmux.2tv\FSharp.ProjectSystem.FSharp.dll&apos; ---&gt; System.Security.SecurityException: Strong name validation failed. (Exception from HRESULT: 0x8013141A)`

**Comment by vasily-kirichenko** [comment]
Ah, you have to disable strong validation checking.

Another thought: can it be the case that your solution has circular project references?

**Comment by vasily-kirichenko** [comment]
`sn -Vr *,*`

**Comment by asik** [comment]
The solution doesn't have circular references. I thought this was impossible but I was curious so I wrote a program to verify it; it really doesn't.

I'm seeing the same behavior using your VSIX.

**Comment by asik** [comment]
Well, on the smaller (61 projects) solution, with your VSIX it seems to more reliably open it and let me use it, after a very long loading time and going well into 2GB of RAM. But the full one still crashes as before.

**Comment by vasily-kirichenko** [comment]
@majocha can it be the case that your "register all projects in workspace as solution is loading" feature can eats all memory and take very long time to finnish?

@asik can you take a dotTrace "timeline" snapshot while it's loading your full solution?

**Comment by majocha** [comment]
@vasily-kirichenko I think it did the same before, just not on load but after opening first file.

**Comment by vasily-kirichenko** [comment]
@majocha no. It loads the project a file belongs to and all referenced projects.

**Comment by saul** [comment]
@vasily-kirichenko that's my concern too. I noticed that there's a call in ProjectSiteAndFiles that is O(2^n) where n is the number of projects in the Solution (with a worst case dependency graph). 

It's high on my list to investigate, but I imagine that memoising that function will significantly improve load time performance. 

My other concern is that Roslyn workspace setup is done async but it context switches back to the UI thread, locking everything up 😔

**Comment by majocha** [comment]
@vasily-kirichenko that was the case for some time but before my change that was a PR that queued all the projects and processed them at once after file open. So it's not entirely my fault :)

**Comment by saul** [comment]
The O(2^n) function I'm referring to is ProjectSitesAndFiles.getProjectOptionsForProjectSite

**Comment by vasily-kirichenko** [comment]
I cannot reproduce it - all my solutions load quite fast. @asik a dotTrace snapshot would be very much appreciated.

**Comment by saul** [comment]
@vasily-kirichenko, this should help you reproduce the issue that @asik is seeing:

1. Download and extract [Perf.zip](https://github.com/Microsoft/visualfsharp/files/672903/Perf.zip​)
2. Run gen.fsx
3. Open `dense/Dense.sln` in Visual Studio

**Comment by vasily-kirichenko** [comment]
@cloudRoutine this thing should be cached https://github.com/Microsoft/visualfsharp/blob/master/vsintegration/src/FSharp.LanguageService/ProjectSitesAndFiles.fs#L111, like
```fsharp
static let solutionService (serviceProvider: System.IServiceProvider) =
    lazy (try Some (serviceProvider.GetService(typeof<SVsSolution>) :?> IVsSolution) with _ -> None)
```

**Comment by cloudRoutine** [comment]
why can't we just pass it the `IVsSolution` as an arg?

**Comment by Pilchie** [comment]
Just wanted to point out that the call to `serviceProvider.GetService()` should effectively be a dictionary lookup in the shell after the first call, so caching it locally in F# isn't likely to help much.

**Comment by asik** [comment]
@saul it's really cool you wrote that. It does repro the long loading time, however not the memory usage going out of control. But hopefully fixing the former would fix the latter. 

I looked at dotTrace and unfortunately it also has file names of all the projects and dlls, can't share that. But if it's certain ETW events you want I can run perfview, any particular command line?

**Comment by forki** [comment]
First PR addressing part of the problem is at https://github.com/Microsoft/visualfsharp/pull/2802

**Comment by dsyme** [comment]
@asik It would be really, really great  if you could manufacture an entirely artificial huge solution that exhibits this problem.  I know it might take a bit of work generating fake projects and code, but if we can capture the solution then we can put it in our automated test cases and ensure that we always scale to a solution of this size.   This would be a hugely appreciated contribution if you could manage it.

**Comment by asik** [comment]
Good news! I can actually repro with @saul 's generated solution, on my work environment. I have no idea why in this environment the memory balloons out of control, whereas at home it doesn't. But maybe at home I'm not running the latest nightly. I'll retest when I get time.

So anyway, dotTrace snapshot from VS slowly dying:
https://www.dropbox.com/s/13swowphpptk8a4/VS2017%20timeline%20snapshot.7z?dl=0

**Comment by vasily-kirichenko** [comment]
@asik could you try https://github.com/Microsoft/visualfsharp/pull/2801, grab vsix here https://drive.google.com/open?id=0B8HLQUKik9VtSlBtVTdfUHhYMjQ?

**Comment by asik** [comment]
So it does load our solution, pretty fast as well. However, after that trying to open files from the solution explorer I am often greeted with
![image](https://cloud.githubusercontent.com/assets/1631157/24765530/2f8735cc-1ac6-11e7-8479-c6b53696097e.png)
and there is no syntax highlighting at all, even on small solutions.

**Comment by saul** [comment]
@asik that is left over from Vasily's investigation. To fix it, create that file on your disk.

**Comment by vasily-kirichenko** [comment]
That VSIX is not usable anyway, it's just POC where the problem is. I'm not gonna provide a fix.

**Comment by dsyme** [comment]
@saul @vasily-kirichenko @cloudRoutine I added the stress test projects generated by @saul directly to the repo in https://github.com/Microsoft/visualfsharp/pull/2997

Checking in the files directly is a bit brutal but we need to have at least a few project stress tests we can share, focus on and automate

**Comment by dsyme** [comment]
Perf info for opening ``tests\projects\stress\medium\dense\Dense.sln``:

Inclusive:
```
Function Name	Inclusive Samples	Exclusive Samples	Inclusive Samples %	Exclusive Samples %
Microsoft.VisualStudio.FSharp.LanguageService.ProjectSitesAndFiles.GetProjectOptionsForProjectSite	24,288	0	58.99	0.00
Microsoft.VisualStudio.FSharp.LanguageService.ProjectSitesAndFiles.getProjectOptionsForProjectSite	24,286	9	58.98	0.02
<StartupCode$FSharp-LanguageService>.$ProjectSitesAndFiles+clo@128-8.Invoke	24,115	12	58.57	0.03
<StartupCode$FSharp-LanguageService>.$ProjectSitesAndFiles+clo@126-7.Invoke	21,896	4	53.18	0.01
Microsoft.VisualStudio.FSharp.LanguageService.ProjectSitesAndFiles.fullOutputAssemblyPath	21,703	23	52.71	0.06
<StartupCode$FSharp-LanguageService>.$ProjectSitesAndFiles.getProperty@87	11,923	7	28.96	0.02
Microsoft.VisualStudio.FSharp.ProjectSystem.Automation.OAProject.get_Properties	9,987	9,987	24.26	24.26
<StartupCode$FSharp-LanguageService>.$ProjectSitesAndFiles+clo@90-1.Invoke	9,568	24	23.24	0.06
```
Exclusive:
```
Function Name	Inclusive Samples	Exclusive Samples	Inclusive Samples %	Exclusive Samples %
Microsoft.VisualStudio.FSharp.ProjectSystem.Automation.OAProject.get_Properties	9,987	9,987	24.26	24.26
[EnvDTE.ni.dll]	8,664	8,449	21.04	20.52
Microsoft.VisualStudio.FSharp.ProjectSystem.ProjectNode.DoMSBuildSubmission	7,000	6,552	17.00	15.91
[mscorlib.ni.dll]	5,870	3,110	14.26	7.55
[Microsoft.VisualStudio.ExtensionEngine.ni.dll]	1,585	1,584	3.85	3.85
[Newtonsoft.Json.ni.dll]	3,442	1,195	8.36	2.90
Microsoft.VisualStudio.FSharp.ProjectSystem.ProjectNode.GetProjectProperty	1,091	1,011	2.65	2.46
Microsoft.VisualStudio.FSharp.ProjectSystem.ProjectNode.Reload	1,634	844	3.97	2.05
Microsoft.VisualStudio.FSharp.ProjectSystem.Automation.OAProject.get_ConfigurationManager	857	818	2.08	1.99
Microsoft.VisualStudio.FSharp.ProjectSystem.Automation.OAProject.get_UniqueName	567	515	1.38	1.25
```

**Comment by dsyme** [comment]
I believe I know how to fix this problem. We will have to change the shape of CompileOptions to no longer recursively specify the dependent projects but rather provide a list of projects options indexed by DLL name.  I will work on a PR for this tomorrow

**Comment by dsyme** [comment]
@saul I've merged #3002 and will close this bug to match, but please submit your run-on-the-background improvements too
